docs/polkadot/implementors-guide/guide.md

```diff
@@ -0,0 +1,544 @@
+
+Note that the candidate can fail to be included in any of the following ways:
+  - The collator is not able to propagate the candidate to any validators assigned to the parachain.
+  - The candidate is not fully backed by validators participating in the Candidate Backing Process.
+  - The candidate is not selected by a relay-chain block author to be included in the relay chain
+  - The candidate's PoV is not considered as available within a timeout and is discarded from the relay chain.
+
+This process can be divided further down. Steps 2 & 3 relate to the work of the collator in collating and distributing the candidate to validators via the Collation Distribution Process. Steps 3 & 4 relate to the work of the validators in the Candidate Backing Process and the block author (itself a validator) to include the block into the relay chain. Steps 6, 7, and 8 correspond to the logic of the relay-chain state-machine (otherwise known as the Runtime) used to fully incorporate the block into the chain. Step 7 requires further work on the validators' parts to participate in the Availability Distribution process and include that information into the relay chain for step 8 to be fully realized.
+
+This brings us to the second part of the process. Once a parablock is considered available and part of the parachain, it is still "pending approval". At this stage in the pipeline, the parablock has been backed by a majority of validators in the group assigned to that parachain, and its data has been guaranteed available by the set of validators as a whole. However, the validators in the parachain-group (known as the "Parachain Validators" for that parachain) are sampled from a validator set which contains some proportion of byzantine, or arbitrarily malicious members. This implies that the Parachain Validators for some parachain may be majority-dishonest, which means that secondary checks must be done on the block before it can be considered approved. This is necessary only because the Parachain Validators for a given parachain are sampled from an overall validator set which is assumed to be up to <1/3 dishonest - meaning that there is a chance to randomly sample Parachain Validators for a parachain that are majority or fully dishonest and can back a candidate wrongly. The Approval Process allows us to detect such misbehavior after-the-fact without allocating more Parachain Validators and reducing the throughput of the system.
+
+The Approval Process looks like this:
+1. Parablocks that have been included by the Inclusion Pipeline are pending approval for a time-window known as the secondary checking window.
+1. During the secondary-checking window, validators randomly self-select to perform secondary checks on the parablock.
```

Comment:
Does "randomly self-select" mean they do it whenever they feel like it? Or they use some verifiable random process? Is there any incentive or mandate to self-select?

Reply:
Yeah, they use a verifiable random process. I assume they are meant to be rewarded for it, but @burdges can shed more light on that.

Reply:
Yes validators use VRFs for this.

If validators self selected their secondary approval checker assignments using only local non-verifiable randomness then well-funded but "overly curious" validators could download all parachains blocks across the network, which (a) reduces our scalability and worse (b) reduces our tricks against adaptive DoSing adversaries.

Instead, all validators have some VRF key with which they compute VRF values that determine their assignments.  If V announces a VRF that says V checks B then (a) V has highest priority for receiving pieces of B, but also (b) we do not proceed into grandpa until either V checks B, or else we declare V a no-show and solicit more previously unknown VRFs.

A strong DoS adversary could force multiple nodes be no-shows until their own nodes fulfilled all required checks.  We expect they'd need to silence a couples times the expected number of checkers though, which increases the odds the attacker hits some hardened honest node with many and/or hidden sentries.  

We hope honest no-shows would be rare because if you're validator was online to gossip out a VRF then they should be there to fetch pieces immediately afterwards, and thus hopefully making them a strong sign of DoS attacks.

Initially we'll need to activate parachains without all this no-show logic because selecting the delay requires real live network experience but that's fine.